zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -547,6 +682,105 @@ impl Chain {
+            "incorrect overwrite of orchard tree: trees must be reverted then inserted",
+        );
+        #[cfg(not(test))]
+        assert_eq!(
+            self.orchard_anchors_by_height.insert(height, anchor),
+            None,
+            "incorrect overwrite of orchard anchor: anchors must be reverted then inserted",
+        );
+
+        #[cfg(test)]
+        self.orchard_trees_by_height.insert(height, tree);
+        #[cfg(test)]
+        self.orchard_anchors_by_height.insert(height, anchor);
+
+        // Multiple inserts are expected here,
+        // because the anchors only change if a block has shielded transactions.
+        self.orchard_anchors.insert(anchor);
+    }
+
+    /// Remove the Orchard tree and anchor indexes at `height`.
+    ///
+    /// `height` can be at two different [`RevertPosition`]s in the chain:
+    /// - a tip block above a chain fork: only that height is removed, or
+    /// - a root block: all trees and anchors below that height are removed,
+    ///   including temporary finalized tip trees.
+    fn remove_orchard_tree_and_anchor(&mut self, position: RevertPosition, height: Height) {
+        let removed_heights: Vec<Height> = if position == RevertPosition::Root {
+            // Remove all trees and anchors at or below the removed block.
+            // This makes sure the temporary trees from finalized tip forks are removed.
+            self.orchard_anchors_by_height
+                .keys()
+                .cloned()
+                .filter(|index_height| *index_height <= height)
+                .collect()
+        } else {
+            // Just remove the reverted tip trees and anchors.
+            vec![height]
+        };
+
+        for height in &removed_heights {
+            let anchor = self
+                .orchard_anchors_by_height
+                .remove(height)
+                .expect("Orchard anchor must be present if block was added to chain");
+            self.orchard_trees_by_height
+                .remove(height)
+                .expect("Orchard note commitment tree must be present if block was added to chain");
+
+            trace!(?height, ?position, ?anchor, "removing orchard tree");
+
+            // Multiple removals are expected here,
+            // because the anchors only change if a block has shielded transactions.
+            assert!(
+                self.orchard_anchors.remove(&anchor),
+                "Orchard anchor must be present if block was added to chain"
+            );
+        }
+    }
+
+    /// Returns the History tree of the tip of this [`Chain`],
+    /// including all finalized blocks, and the non-finalized blocks below the chain tip.
+    ///
+    /// If the chain is empty, instead returns the tree of the finalized tip,
+    /// which was supplied in [`Chain::new()`]
+    ///
+    /// # Panics
+    ///
+    /// If this chain has no history trees. (This should be impossible.)
+    pub fn history_tree_at_tip(&self) -> Arc<HistoryTree> {
```

Comment:
Nitpick: rename this to `tip_history_tree` for consistency with the other method names.

Reply:
I think the most consistent name might be `history_block_commitment_tree`, to match `{sprout,sapling,orchard}_note_commitment_tree`?

We can do more renames of existing functions/fields later, I was hoping to keep this PR small.

Reply:
Changed in 0e2335e836